stacks-signer/src/stacks_client.rs

```diff
@@ -98,6 +172,173 @@ impl StacksClient {
+            }
+        };
+        let mut unsigned_tx = StacksTransaction::new(self.tx_version, auth, payload);
+        unsigned_tx.anchor_mode = anchor_mode;
+        unsigned_tx.post_condition_mode = TransactionPostConditionMode::Allow;
+        unsigned_tx.chain_id = self.chain_id;
+
+        let mut tx_signer = StacksTransactionSigner::new(&unsigned_tx);
+        tx_signer
+            .sign_origin(&self.stacks_private_key)
+            .map_err(|_| ClientError::SignatureGenerationFailure)?;
+        if let (Some(payer), Some(_)) = (payer, payer_nonce) {
+            tx_signer
+                .sign_sponsor(payer)
+                .map_err(|_| ClientError::SponsorSignatureGenerationFailure)?;
+        }
+
+        let Some(tx) = tx_signer.get_tx() else {
+            return Err(ClientError::SignatureGenerationFailure);
+        };
+
+        Ok(tx.serialize_to_vec())
+    }
+}
+
+impl StacksContractCallable for StacksClient {
+    fn transaction_contract_call(
+        &self,
+        nonce: u64,
+        contract_addr: &StacksAddress,
+        contract_name: &str,
+        function_name: &str,
+        function_args: &[ClarityValue],
+    ) -> Result<Vec<u8>, ClientError> {
+        let contract_name = ContractName::from(contract_name);
+        let function_name = ClarityName::from(function_name);
+
+        let payload = TransactionContractCall {
+            address: *contract_addr,
+            contract_name,
+            function_name,
+            function_args: function_args.to_vec(),
+        };
+
+        let tx_fee = 0;
+
+        self.serialize_sign_sig_tx_anchor_mode_version(
+            payload.into(),
+            nonce,
+            tx_fee,
+            TransactionAnchorMode::OnChainOnly,
+        )
+    }
+
+    fn submit_tx(&self, tx: Vec<u8>) -> Result<String, ClientError> {
+        let path = format!("{}/v2/transactions", self.http_origin);
+        let res = self
+            .stacks_node_client
+            .post(path)
+            .header("Content-Type", "application/octet-stream")
+            .body(tx.clone())
+            .send()
+            .unwrap();
+        if res.status().is_success() {
+            let res: String = res.json().unwrap();
+            assert_eq!(
+                res,
+                StacksTransaction::consensus_deserialize(&mut &tx[..])
+                    .unwrap()
+                    .txid()
+                    .to_string()
+            );
+            Ok(res)
+        } else {
+            Err(ClientError::TransactionSubmissionFailure)
+        }
+    }
+
+    fn read_only_contract_call(
+        &self,
+        contract_addr: &StacksAddress,
+        contract_name: &str,
+        function_name: &str,
+        function_args: &[&str],
+    ) -> Result<String, ClientError> {
+        debug!("Calling read-only function {}...", function_name);
+        let body = json!({"sender": self.stacks_address.to_string(), "arguments": function_args})
```

Comment:
This can fail if `function_args` does not contain serialized `clarity::vm::Value`s.  This interface should instead take `&[clarity::vm::Value]` for `function_args`.